src/core/Database.h

```diff
@@ -99,76 +95,99 @@ class Database : public QObject
     QSharedPointer<const CompositeKey> key() const;
     QByteArray challengeResponseKey() const;
     bool challengeMasterSeed(const QByteArray& masterSeed);
 
-    void setCipher(const QUuid& cipher);
-    void setCompressionAlgo(Database::CompressionAlgorithm algo);
-    void setKdf(QSharedPointer<Kdf> kdf);
-    bool setKey(const QSharedPointer<const CompositeKey>& key, bool updateChangedTime = true, bool updateTransformSalt = false);
-    bool hasKey() const;
-    bool verifyKey(const QSharedPointer<CompositeKey>& key) const;
-    QVariantMap& publicCustomData();
-    const QVariantMap& publicCustomData() const;
-    void setPublicCustomData(const QVariantMap& customData);
     void recycleEntry(Entry* entry);
     void recycleGroup(Group* group);
     void emptyRecycleBin();
+
+    bool isModified() const;
     void setEmitModified(bool value);
-    void markAsModified();
-    QString saveToFile(const QString& filePath, bool atomic = true, bool backup = false);
 
-    /**
-     * Returns a unique id that is only valid as long as the Database exists.
-     */
-    const QUuid& uuid();
-    bool changeKdf(const QSharedPointer<Kdf>& kdf);
+    Group* resolveGroup(const QUuid& uuid);
+    Entry* resolveEntry(const QUuid& uuid);
+    Entry* resolveEntry(const QString& text, EntryReferenceType referenceType);
 
-    static Database* databaseByUuid(const QUuid& uuid);
-    static Database* openDatabaseFile(const QString& fileName, QSharedPointer<const CompositeKey> key);
-    static Database* unlockFromStdin(const QString& databaseFilename, const QString& keyFilename = {},
-        FILE* outputDescriptor = stdout, FILE* errorDescriptor = stderr);
+    static Database* databaseByUuid(const QUuid& uuid);;
+    static Database* databaseByFilePath(const QString& filePath);;
+    static QSharedPointer<Database> unlockFromStdin(QString databaseFilename, QString keyFilename = {},
+                                                    FILE* outputDescriptor = stdout, FILE* errorDescriptor = stderr);
+
+public slots:
+    void markAsModified();
+    void markAsClean();
 
 signals:
+    void filePathChanged(const QString& oldPath, const QString& newPath);
     void groupDataChanged(Group* group);
     void groupAboutToAdd(Group* group, int index);
     void groupAdded();
     void groupAboutToRemove(Group* group);
     void groupRemoved();
     void groupAboutToMove(Group* group, Group* toGroup, int index);
     void groupMoved();
-    void nameTextChanged();
     void modified();
-    void modifiedImmediate();
+    void metadataModified();
+    void clean();
 
 private slots:
     void startModifiedTimer();
 
 private:
+    struct DatabaseData
+    {
+        QString filePath;
+        bool isReadOnly = false;
+        QUuid cipher = KeePass2::CIPHER_AES256;
+        CompressionAlgorithm compressionAlgo = CompressionGZip;
+        QByteArray transformedMasterKey;
+        QSharedPointer<Kdf> kdf = QSharedPointer<AesKdf>::create(true);
```

Comment:
Shouldn't this be default to Argon2 Kdf?

Reply:
No. This is a left-over of our previous KDBX3-centric architecture where we needed KDBX3 to be the default. We still do for our tests (and perhaps other stuff). Changing this is a huge mess. If the DB is KDBX4, the KDF will be re-initialised anyway, whereas KDBX3 does not do any KDF initialisation, hence this is needed.